### PR TITLE
Add hi and goodbye

### PR DESCRIPTION
**Pull request by binaryoverload**
# Pull request

 - [x] I have read and agreed to the [code of conduct](https://github.com/CascadeBot/CascadeBot/blob/master/.github/CODE_OF_CONDUCT.md).
 - [x] I have read and complied with the [contributing guidelines](https://github.com/CascadeBot/CascadeBot/blob/master/.github/CONTRIBUTING.md).
 - [x] What I'm implementing was assigned to me or was not being worked on and is a somewhat requested feature. For reference see our [GitHub projects](https://github.com/orgs/CascadeBot/projects).
 - [x] I have tested all of my changes.
 
So I think we should have hello and goodbye messages that are sent when we are added/removed from a guild. This will allow us to include things like: 
 - Links to our panel
 - A removal survey
 - Awareness on things people need to be aware of
 - Etc etc

### CODE REVIEW

File: config.example.yml [hunk 1]
```diff
@@ -19,6 +19,17 @@
...
+# This is the message that will either be DMed to the guild owner
+# or posted in the first text channel whichever comes first.
+guild_welcome: |-
+  Thank you for adding CascadeBot!
+
+# This is the message that will be DMed to the guild owner
+# when we are removed from a guild.
+guild_goodbye: |-
+  We're sorry that this is goodbye! We hope we see you again soon :(
+  If you have a moment, please let us know what we did wrong here: <insert form link>
```
Comment (JamesHawkinss) [inline config.example.yml]: I suggest changing your wording here. We might not have done something wrong, perhaps say something like "please let us know why this is goodbye here:"

**Comment by binaryoverload** [comment]
This is just an example, we'll come up with a lot better one in future. And imo we should always act if we have done something wrong from a marketing angle (Because if we've lost a server, we've failed 😂 )

File: src/main/java/com/cascadebot/cascadebot/events/GeneralEvents.java [hunk 2]
```diff
@@ -20,4 +25,27 @@ public void onReady(ReadyEvent event) {
...
     }
 
+    @Override
+    public void onGuildJoin(GuildJoinEvent event) {
+        if (!StringUtils.isBlank(Config.INS.getGuildWelcomeMessage())) {
+            Guild guild = event.getGuild();
+            guild.getOwner().getUser().openPrivateChannel().queue(privateChannel -> {
+                privateChannel.sendMessage(Config.INS.getGuildWelcomeMessage()).queue();
+            }, error -> {
+                if (guild.getDefaultChannel() != null) {
```
Comment (weeryan17) [inline src/main/java/com/cascadebot/cascadebot/events/GeneralEvents.java]: Aren't default channels not a thing any more? also we should send in a channel at all if the owners dms are closed imo.

**Comment by mrjvs** [comment]
Im against the idea of using DM's to convey these types of things. if people leave, so be it. they dont want to be pestered by surveys or messages. they just want to leave since they are done investigating the server. having a welcome message in a text channel to show the rules and welcome them is something I like tho.

File: src/main/java/com/cascadebot/cascadebot/events/GeneralEvents.java (hunk 2 above)
Comment (binaryoverload) [reply src/main/java/com/cascadebot/cascadebot/events/GeneralEvents.java]: In JDA a default channel is defined as:
```
This channel is the first channel in the guild (ordered by position) that the getPublicRole() has the Permission.MESSAGE_READ in.
```
Good point about whether we should or not 🤔

**Comment by JamesHawkinss** [comment]
> Im against the idea of using DM's to convey these types of things. if people leave, so be it. they dont want to be pestered by surveys or messages. they just want to leave since they are done investigating the server. having a welcome message in a text channel to show the rules and welcome them is something I like tho.

This is for when the bot is removed or added from a guild, not when a person leaves or joins

**Comment by mrjvs** [comment]
> 
> 
> > Im against the idea of using DM's to convey these types of things. if people leave, so be it. they dont want to be pestered by surveys or messages. they just want to leave since they are done investigating the server. having a welcome message in a text channel to show the rules and welcome them is something I like tho.
> 
> This is for when the bot is removed or added from a guild, not when a person leaves or joins

erhm, shush. I knew that 😓

**Comment by binaryoverload** [comment]
@mrjvs Opinion on sending in the first available channel if DMs not available?

**Comment by mrjvs** [comment]
> 
> 
> @mrjvs Opinion on sending in the first available channel if DMs not available?

dont, dm's are disabled for a reason. and on big servers you dont want bots to randomly send messages.

**Comment by JamesHawkinss** [comment]
> @mrjvs Opinion on sending in the first available channel if DMs not available?

@binaryoverload I know I wasn't asked but I think it should just DM, not send to first available channel

**Comment by Ikinon** [comment]
@binaryoverload
I know I was not involved but:
What if we searched for a channel named something like "bot-channel"
If channel don't exist just leave it?

**Comment by JamesHawkinss** [comment]
> @binaryoverload
> I know I was not involved but:
> What if we searched for a channel named something like "bot-channel"
> If channel don't exist just leave it?

We would have to do a lot of research to effectively put this in place, because loads of people call it loads of different things

**Comment by Ikinon** [comment]
True

**Comment by binaryoverload** [comment]
> @binaryoverload
> I know I was not involved but:
> What if we searched for a channel named something like "bot-channel"
> If channel don't exist just leave it?

We could do that but like @JamesHawkinss said people call them loads of different things. I'll just remove it :D

**Pull request merged**